CRM/Mailing/BAO/Mailing.php

```diff
@@ -109,46 +109,227 @@ public static function getRecipientsCount($job_id, $mailing_id = NULL) {
+                    SELECT      $email.id as email_id, c.id as contact_id
+                    FROM   $email
+                    INNER JOIN  civicrm_contact c ON $email.contact_id = c.id
+                    INNER JOIN  civicrm_group_contact gc ON gc.contact_id = c.id
+                    INNER JOIN  civicrm_mailing_group mg  ON  gc.group_id = mg.entity_id AND mg.search_id IS NULL
+                    LEFT JOIN   %s temp ON c.id = temp.contact_id
+                    WHERE gc.group_id IN (%s) AND gc.status = 'Added' AND %s
+                    GROUP BY $email.id, c.id
+                    %s", $includedTempTablename, $excludeTempTablename, implode(',', $recipientsGroup['Include']), implode(' AND ', $includeFilters), $order_by);
+    $mailingGroup->query($sql);
+
+    if (count($includeSmartGroupIDs)) {
+      unset($includeFilters["gc.status = 'Added'"]);
+      $sql = sprintf("REPLACE INTO %s (email_id, contact_id)
+      SELECT  civicrm_email.id as email_id, c.id as contact_id
+      FROM       civicrm_contact c
+      INNER JOIN civicrm_email ON civicrm_email.contact_id = c.id
+      INNER JOIN civicrm_group_contact_cache gc ON gc.contact_id = c.id
+      LEFT  JOIN %s temp              ON temp.contact_id = c.id
+      WHERE      gc.group_id IN (%s) AND %s %s ", $includedTempTablename, $excludeTempTablename, implode(',', $includeSmartGroupIDs), implode(' AND ', $includeFilters), $order_by);
+      $mailingGroup->query($sql);
     }
-    $contact = CRM_Contact_DAO_Contact::getTableName();
 
-    $group = CRM_Contact_DAO_Group::getTableName();
-    $g2contact = CRM_Contact_DAO_GroupContact::getTableName();
+    // Construct the filtered search queries.
+    $query = " SELECT search_id, search_args, entity_id
+      FROM   civicrm_mailing_group
+      WHERE  search_id IS NOT NULL AND mailing_id = {$mailingID}";
+    $dao = CRM_Core_DAO::executeQuery($query);
+    while ($dao->fetch()) {
+      $customSQL = CRM_Contact_BAO_SearchCustom::civiMailSQL($dao->search_id,
+        $dao->search_args,
+        $dao->entity_id
+      );
+      $query = "REPLACE INTO {$includedTempTablename} (email_id, contact_id) {$customSQL} ";
+      $mailingGroup->query($query);
+    }
 
-    $m = new CRM_Mailing_DAO_Mailing();
-    $m->id = $mailing_id;
-    $m->find(TRUE);
+    list($aclFrom, $aclWhere) = CRM_Contact_BAO_Contact_Permission::cacheClause();
+    $aclWhere = $aclWhere ? "WHERE {$aclWhere}" : '';
+
+    // clear all the mailing recipients before populating
+    $sql = " DELETE FROM civicrm_mailing_recipients WHERE  mailing_id = %1 ";
+    CRM_Core_DAO::executeQuery($sql, array(1 => array($mailingID, 'Integer')));
+
+    $selectClause = array('%1', 'i.contact_id', "i.email_id");
+    $select = "SELECT " . implode(', ', $selectClause);
+    // CRM-3975
+    $groupBy = $groupJoin = '';
+    $orderBy = "i.contact_id, i.email_id";
+    if ($mailingObj->dedupe_email) {
+      $orderBy = "MIN(i.contact_id), MIN(i.email_id)";
```

Comment:
I think DISTINCT, a synonym for DISTINCTROW, is more efficient than doing an order by.

Reply:
old cruft code :( will change